### PR TITLE
Update style guide

### PR DESCRIPTION
**Pull request by sipa**
As discussed in this week's IRC meeting.

### CODE REVIEW

**Review by dcousens — APPROVED** [review]

**Review by paveljanik — APPROVED** [review]

**Comment by practicalswift** [comment]
ACK 47d8441

Very good! Thanks for clarifying!

File: doc/developer-notes.md [hunk 1]
```diff
@@ -3,41 +3,64 @@ Developer Notes
...
+- **Symbol naming conventions**. These are preferred in new code, but are not
+required when doing so would need changes to significant pieces of existing
+code.
+  - Variable and namespace names are all lowercase, and may use `_` to
+    separate words.
+    - Class member variables have a `m_` prefix.
+    - Global variables have a `g_` prefix.
+  - Constant names are all uppercase, and use `_` to separate words.
+  - Class names, function names and method names are CamelCase. Do not prefix
+    class names with `C`.
```
Comment (luke-jr) [inline doc/developer-notes.md]: Why not prefix class names with `C`?

**Comment by sipa** [comment]
It is always obvious from the usage that a name refers to a class and not a function/method, so no need to repeat that information in the name.

**Comment by luke-jr** [comment]
Is it? `function(0, class())` seems fairly non-obvious.

**Comment by gmaxwell** [comment]
Seems fine to me.

File: doc/developer-notes.md [hunk 2]
```diff
@@ -3,41 +3,64 @@ Developer Notes
...
+
+- **Symbol naming conventions**. These are preferred in new code, but are not
+required when doing so would need changes to significant pieces of existing
+code.
+  - Variable and namespace names are all lowercase, and may use `_` to
+    separate words.
+    - Class member variables have a `m_` prefix.
+    - Global variables have a `g_` prefix.
+  - Constant names are all uppercase, and use `_` to separate words.
+  - Class names, function names and method names are CamelCase. Do not prefix
```
Comment (jonasschnelli) [inline doc/developer-notes.md]: Not sure if we had a discussion about that already:
I often saw projects where the static class methods and global space functions where in `UpperCamelCase()` while the class methods where in `lowerCamelCase()` and I think this is useful.
But no strong opinion.

**Comment by jonasschnelli** [comment]
ACK 47d84414662c09bc08cfc96f78457ddd602a20ac

File: doc/developer-notes.md (hunk 2 above)
Comment (laanwj) [inline doc/developer-notes.md]: Except for the qt code, this just documents the style already in use.
So while this sounds sensible for a new project, I don't think we should change this now.

File: doc/developer-notes.md [hunk 3]
```diff
@@ -3,41 +3,64 @@ Developer Notes
...
 and the result is not very consistent. However, we're now trying to converge to
-a single style, so please use it in new code. Old code will be converted
-gradually and you are encouraged to use the provided
-[clang-format-diff script](/contrib/devtools/README.md#clang-format-diffpy)
-to clean up the patch automatically before submitting a pull request.
+a single style, which is specified below. When writing patches, favor the new
+style over attempting to mimick the surrounding style, except for move-only
+commits.
+
+Do not submit patches solely to modify the style of existing code.
```
Comment (laanwj) [inline doc/developer-notes.md]: Thanks for adding this explicitly :+1:

**Comment by laanwj** [comment]
utACK 47d8441

**Comment by jtimon** [comment]
ACK

**Pull request merged**